### PR TITLE
Support Centipede testcase minimization

### PR DESCRIPTION
**Pull request by Alan32Liu**
1. Support testcase minimization feature for Centipede.
2. Add a unittest for testcase minimization for Centipede.
3. Update Centipede binary to allow new features
4. Add a testcase and a fuzz target binary for the unittest of testcase minimization.

### CODE REVIEW

**Comment by Alan32Liu** [comment]
Will submit another PR to support minimizing corpus.

File: src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py [hunk 1]
```diff
@@ -309,6 +310,28 @@ def minimize_corpus(self, target_path, arguments, input_dirs, output_dir,
     """
     raise NotImplementedError
 
+  def _get_smallest_crasher(self, workdir_path):
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py:313]: How long is this test btw?

**Review by jonathanmetzman — APPROVED** [review]
lgtm

File: src/clusterfuzz/_internal/tests/core/bot/fuzzers/centipede/test_data/minimize_me_fuzz_target.cc [hunk 2]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2022 The Centipede Authors.
```
[the remark this answers is not in the document]
Comment (Alan32Liu) [reply src/clusterfuzz/_internal/tests/core/bot/fuzzers/centipede/test_data/minimize_me_fuzz_target.cc:1]: It's directly copied from [`Centipede`'s repo](https://github.com/google/fuzztest/blob/390fc71df5f2dfbb588a6d572121e8c7e61f9c5f/centipede/testing/minimize_me_fuzz_target.cc).
Do we want to change it?

File: src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py (hunk 1 above)
Comment (Alan32Liu) [reply src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py:313]: This new test takes ~0.17s.
All 25 tests take ~22.162s in total, the slowest two are `test_crash_timeout`s from `UnshareIntegrationTest` and `IntegrationTest`, taking ~8.09s each.

**Comment by Alan32Liu** [comment]
Force-pushed code to resolve conflicts with #3419

File: src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py [hunk 3]
```diff
@@ -302,6 +302,28 @@ def minimize_corpus(self, target_path, arguments, input_dirs, output_dir,
...
+        if os.path.isfile(os.path.join(crashes_path, t))
+    ]
+    if not testcases:
+      logs.log_error(f'No crash testcases under: {crashes_path}')
+      return None
+
+    minimum_testcase = min(testcases, key=os.path.getsize)
+    return minimum_testcase
+
   def minimize_testcase(self, target_path, arguments, input_path, output_path,
```
Comment (Alan32Liu) [inline src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py:327]: Hi @lszekeres, this PR supports testcase minimization for `Centipede`: The `input_path` takes the testcase to be minimized, and `output_path` points to the same testcase after minimization.
For example, the unittest minimizes `?f???u???z?` to `?fuz?` while maintaining its ability to reproduce the crash.

Does this match the feature you requested via email earlier?
Please let me know if I missed anything, thanks : )
I will push another PR for corpus minimization.

Comment (lszekeres) [reply src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py:327]: LGTM, thanks!

File: src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py [hunk 4]
```diff
@@ -316,7 +338,24 @@ def minimize_testcase(self, target_path, arguments, input_path, output_path,
...
+    args = [
+        f'--binary={target_path}',
+        f'--workdir={workdir}',
+        f'--minimize_crash={input_path}',
+        f'--num_runs={constants.NUM_RUNS_PER_MINIMIZATION}',
+        '--seed=1',
+    ]
+    result = runner.run_and_wait(additional_args=args, timeout=max_time)
+    if result.timed_out:
+      logs.log_error(
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py:352]: We should be consistent here and follow libfuzzer's approach of raising a timeouterror isntead: https://github.com/google/clusterfuzz/blob/7325e7d2d7e4a30ce2a599cbb1f1e5cceac8643c/src/clusterfuzz/_internal/bot/fuzzers/libFuzzer/engine.py#L567

File: src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py [hunk 5]
```diff
@@ -316,7 +338,24 @@ def minimize_testcase(self, target_path, arguments, input_path, output_path,
...
+          'Testcase minimization timed out.', fuzzer_output=result.output)
+    minimum_testcase = self._get_smallest_crasher(workdir)
+    if minimum_testcase:
+      shutil.copyfile(minimum_testcase, output_path)
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py:356]: We should fall back to just copying the original testcase as is to the output_path as well, if minimization failed here somehow.

**Review by oliverchang — COMMENTED** [review]
nice!

File: src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py (hunk 5 above)
Comment (Alan32Liu) [reply src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py:357]: Fixed.

File: src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py (hunk 4 above)
Comment (Alan32Liu) [reply src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py:352]: Fixed.

**Pull request merged**